app/routes/result.js

```diff
@@ -6,31 +6,57 @@
 import { hash } from 'rsvp';
 import Route from '@ember/routing/route';
 import { inject as service } from '@ember/service';
+import Queryable from 'bullet-ui/mixins/queryable';
 
-export default Route.extend({
+export default Route.extend(Queryable, {
   querier: service(),
+  queryManager: service(),
 
   model(params) {
     return this.store.findRecord('result', params.result_id).catch(() => {
       this.transitionTo('missing', 'not-found');
     });
   },
-  // Force the fetching of query and filter
+
   afterModel(model) {
-    return hash({
-      segments: model.get('segments'),
-      query: model.get('query'),
-      filter: model.get('query').then(query => query.get('filter'))
-    });
+    // Fetch all the things
+    return model.get('query').then(query => {
+      return hash({
+        filter: query.get('filter'),
+        projections: query.get('projections'),
+        aggregation: query.get('aggregation').then(aggregation => {
+          return hash({ groups: aggregation.get('groups'), metrics: aggregation.get('metrics') });
+        }),
+        window: query.get('window')
+      });
+    })
   },
 
   actions: {
     queryClick(query) {
       this.transitionTo('query', query.get('id'));
     },
 
-    willTransition() {
+    reRunClick(query) {
+      this.get('queryManager').addResult(query.get('id')).then(result => {
+        this.get('querier').cancel();
+        // Sends us to a new result page but don't want to cancel the new query when we transition.
+        this.set('skipCancelling', true);
```

Comment:
This may cause query not cancelled and leave websocket connection open forever if leaving the result page immediately after clicking rerun button. We may think about a way to do transition first and then save pending client.

Reply:
Done and added tests